lib/puppet/util/windows/monkey_patches/dir.rb

```diff
@@ -0,0 +1,40 @@
+require 'win32/dir/constants'
+require 'win32/dir/functions'
+require 'win32/dir/structs'
+
+class DirMonkeyPatched
+  include ::Dir::Structs
+  include ::Dir::Constants
+  extend  ::Dir::Functions
+
+  path  = nil
+  key   = 'PERSONAL'
+  value = 0x0005
+  buf   = 0.chr * 1024
+  buf.encode!('UTF-16LE')
+
+  if SHGetFolderPathW(0, value, 0, 0, buf) == 0 # Current path
+    path = buf.strip
+  elsif SHGetFolderPathW(0, value, 0, 1, buf) == 0 # Default path
+    path = buf.strip
+  else
+    ptr   = FFI::MemoryPointer.new(:long)
+    info  = SHFILEINFO.new
+    flags = SHGFI_DISPLAYNAME | SHGFI_PIDL
+
+    if SHGetFolderLocation(0, value, 0, 0, ptr) == 0
+      if SHGetFileInfo(ptr.read_long, 0, info, info.size, flags) != 0
+        path = info[:szDisplayName].to_s
+      end
+    end
+  end
+
+  if path.nil?
+    path = ''
+    begin
+      Dir.const_set(key, path.encode(Encoding.default_external))
+    rescue Encoding::UndefinedConversionError
+      Dir.const_set(key, path.encode('UTF-8'))
```

Comment:
Would just do `Dir.const_set(key, ''.encode...` here without setting the `path` variable, but that's just my preference. Feel free to dismiss this if you want.